Feature/broadcast edits

arminWel opened the pull request:
This PR implements the user stories Broadcast and Apply Edits. It is unit tested, but could not yet be integration tested, as my paranoia arch needs some further setup for this. 

There is some potential for further refactoring, but that would be on a greater scheme and i would suggest a dedicated branch therefore.

ADimeo committed:
Rename methods from msg to message

ADimeo committed:
Add basic test coverage for ClientManager

This includes tests for braodcast messages, and a refactor of how we structure our mocks.

ADimeo committed:
Reorganize methods

arminWel committed:
Merge branch 'develop' into feature/broadcastEdits

arminWel committed:
Allow for MethodAddedOrChangedEdit s to be applied

arminWel committed:
Add classes for all CCEMethodEdits

arminWel committed:
Add tests for CCEEdit and CCEMethodAddedOrModifiedEdit

arminWel committed:
Refactor PackageManagerTests

arminWel committed:
Allow Recategorized-Edits to be applied

arminWel committed:
Fix wrong refrecences in CCEPackageManagerTests

After refactoring of CCEPackageManager there were some deprecated references denying the tests to suceed.

arminWel committed:
Allow for removed-Edits to be applied

arminWel committed:
Refactor in order to apply edits according to idioms

arminWel committed:
Let the packageManger ignore specified edits

arminWel committed:
Fix bugs, which appeared in the last commit

arminWel committed:
Add tests for ignoring packages

ADimeo commented on packages/CollabCodeEdit-Client.package/CCEPackageManager.class/instance/noMoreIgnoreEdit..st:
```diff
@@ -0,0 +1,3 @@
+accessing
+noMoreIgnoreEdit: aCCEEdit
```
Why isn't this called "stopIgnoringEdit"

ADimeo commented on packages/CollabCodeEdit-Core.package/CCEEdit.class/README.md:
```diff
@@ -0,0 +1,11 @@
+	editContent:		Object
+
+classIdentifier
+	- The classIdentifier is the name of the class edited.
```
Minor quibble: If the title is "classIdentifiert" and the first two words are "the classIdentifier is the" I do not get anything of value. Please rephrase, perhaps to something like "The name of the edited class".

ADimeo commented on packages/CollabCodeEdit-Core.package/CCEEdit.class/README.md:
```diff
@@ -0,0 +1,11 @@
+	- The classIdentifier is the name of the class edited.
+
+editContent
+	- The editcontent is an abstract variable holding the Information necessary for applying myself. It is specified by the subclasses.
```
Same here.

ADimeo commented on packages/CollabCodeEdit-Core.package/CCEEdit.class/instance/securelyApplyin..st:
```diff
@@ -0,0 +1,8 @@
+applying
+securelyApplyin: aCCESession
```
Please rename to securelyApplyIn (Uppercase I)

ADimeo commented on packages/CollabCodeEdit-Core.package/CCEEdit.class/class/newWith.in..st:
```diff
@@ -0,0 +1,5 @@
+instance creation
+newWith: anObject in: aClass
```
Input unspecified, please rename (aMethodSource?)

arminWel replied:
This is intentionally `anObject`, as the type of `editContent` is not specified by `CCEEdit`, but rather by it's subclasses

simonstadlinger replied:
Then i would go for the first all possibilities covering class, which would be `CCEEdit` right?

arminWel replied:
No it is not. The `editcontent` is a variable in a `CCEEdit`, therefore it should never be a `CCEEdit`.

ADimeo commented on packages/CollabCodeEdit-Core.package/CCEMessage.class/class/forwardEditMessage..st:
```diff
@@ -1,4 +1,4 @@
 server-Message creation
-forwardEditMessage: aSerializeddCCEEdit
+forwardEditMessage: aCCEEdit
 	^ self new: (CCEMessage serverMsgTypes at: #receiveEdit:) 
```
Please rename serverMsgTypes to a squeak-style name, perhaps serverMessageTypes?

PaulBrachmann replied:
2nd

ADimeo commented on packages/CollabCodeEdit-Core.package/CCEMessage.class/class/recieveEdit..st:
```diff
@@ -0,0 +1,4 @@
+server-Message creation
+recieveEdit: aCCEEdit
```
Please rename to receiveEdit

PaulBrachmann replied:
Yes, please.

ADimeo commented on packages/CollabCodeEdit-Core.package/CCEMethodAddedOrModifiedEdit.class/README.md:
```diff
@@ -0,0 +1,9 @@
+	editContent: Text
+	
+editContent
+	- The editContent is the sourcecode of the changed method
```
Same as above

ADimeo commented on packages/CollabCodeEdit-Core.package/CCEMethodRecategorizedEdit.class/README.md:
```diff
@@ -0,0 +1,9 @@
+	editContent: OrderedDictionary
+	
+editContent
+	- The editContent is a dictionary holding the new category (at self categorySymbol) and the method which is recategorized (at self methodSymbol). Those to may be used like standard instance variables 
```
Same as above.

ADimeo commented on packages/CollabCodeEdit-Core.package/CCEMethodRemovedEdit.class/README.md:
```diff
@@ -0,0 +1,9 @@
+	editContent: Symbol
+	
+editContent
+	- The editContent is the selector of the method removed.
```
Same as above. Additionally, it is not trivial what the selector of the method is. A sentence or two detailing what a selector is might improve maintainability.

ADimeo commented on packages/CollabCodeEdit-Tests.package/CCEClientMock.class/README.md:
```diff
@@ -0,0 +1,9 @@
+A CCEClientMock is used to test integration between a CCEClientManager and its clients.
```
Please use first person

ADimeo commented on packages/CollabCodeEdit-Tests.package/CCEClientMock.class/README.md:
```diff
@@ -0,0 +1,9 @@
+	changesDict:		<Object>
+
+changesDict
+	- xxxxx
```
Please remove placeholder

ADimeo commented on packages/CollabCodeEdit-Tests.package/CCEMock.class/README.md:
```diff
@@ -0,0 +1,9 @@
+A CCEMock is a superclass for all mocks.
+
+It collects messages. A message does not correspond to a CCEMessage, it is anytthing that might be sent to the mock, and might be tested for.
```
Please use first person

ADimeo commented on packages/CollabCodeEdit-Tests.package/CCEMock.class/README.md:
```diff
@@ -0,0 +1,9 @@
+	messages:		<Object>
+
+messages
+	- xxxxx
```
please remove placeholder

ADimeo commented on packages/CollabCodeEdit-Tests.package/CCEMock.class/README.md:
```diff
@@ -0,0 +1,9 @@
+It collects messages. A message does not correspond to a CCEMessage, it is anytthing that might be sent to the mock, and might be tested for.
+
+Instance Variables
+	messages:		<Object>
```
please remove placeholder

ADimeo commented on packages/CollabCodeEdit-Tests.package/CCEPackageManagerTests.class/instance/testNoMoreIgnoreEdit.st:
```diff
@@ -0,0 +1,9 @@
+tests
+testNoMoreIgnoreEdit
```
With renaming of the original method this should also be renamed

ADimeo committed:
Minor style fixes

ADimeo committed:
Minor modification of class comments

arminWel committed:
Merge branch 'develop' into feature/broadcastEdits

arminWel committed:
Clean up after merging

arminWel committed:
Fix incomplete change of names

All method which include `msg` were changed to `message` but not all messages were updated.

simonstadlinger committed:
Merge fix of branch 'develop' of origin into feature/broadcastEdits

simonstadlinger commented on packages/CollabCodeEdit-Client.package/CCEPackageManager.class/instance/listClassesOfScope.st:
```diff
@@ -3,5 +3,5 @@ listClassesOfScope
 	classes := OrderedCollection new.
-	self scope do: [ :package | classes addAll: (self listClassesOfPackage: package) ].
+	session scope do: [ :package | classes addAll: (self listClassesOfPackage: package) ].
 	^ classes
```
Here you could use the collect message of the collection protocol to simplify the code
`^ session scope collect: [ :package | self listClassesOfPackage: package ] `

arminWel replied:
Sadly i cannot do that, as it would result in an array of array, which would overcomplicate the code at other points.

simonstadlinger commented on packages/CollabCodeEdit-Client.package/CCEPackageManager.class/instance/listenForChangesFor..st:
```diff
@@ -0,0 +1,5 @@
+change-capturing
+listenForChangesFor: aCCESession
+	self
```
you can put `self session: ` in one line

arminWel replied:
I could. But when using cascades, this is considered to obscure the intention of the code. Such, it was an intentionally made decision.

PaulBrachmann replied:
I agree with Armin here

simonstadlinger commented on packages/CollabCodeEdit-Client.package/CCEPackageManager.class/instance/listenForMethodChanges.st:
```diff
@@ -1,8 +1,10 @@
+		notify: self ofSystemChangesOfItem: #method change: #Added 
+			using: #methodAddedOrModified:;
+		notify: self ofSystemChangesOfItem: #method change: #Modified 
+			using: #methodAddedOrModified:;
```
verry nice formatting!

simonstadlinger commented on packages/CollabCodeEdit-Client.package/CCEPackageManager.class/instance/methodMoved..st:
```diff
@@ -4,5 +4,5 @@ methodMoved: anEvent
 	(self isInScope: class)
-		ifTrue: [ self session broadcastMovedMethod: anEvent itemSelector 
+		ifTrue: [ self session broadcastRecategorizedMethod: anEvent itemSelector 
 			in: class into: anEvent itemProtocol ]
```
i would suggest to rename The method to `methodRecategorized` in order to stay consistent with your established naming convention for the method edits. Also this method boils down to the `broadcastRecategorizedMethod` which with the right method name appears to be expected as in the other two methods of that kind

simonstadlinger replied:
What i also want to ask is, why did you gave the responsability of `broadcastRecategorizedMethod:` and all other broadcastMethods to the CCESession? In my opinion this violates the separation of concerns pattern. The fact that within the broadcast methods your only reference is the package managers - in checking if the edit should be ignored - , hardens my wondering, why don't put everything concerning edits into the package manager class. We can also discuss why not to put the edit selection receiving applying responsability stuff into a completely new class and let this class reference to the package manager just to check whether the edit captured is in the package scope or not. 

Lets discuss that please. Or tell me if i'm missing an important detail here.

simonstadlinger replied:
As the understanding of your implementation goes further i would highly suggest to separate this concern into a new class. the only thing you do in the session, when calling a method that relates to edits, is to forward the request to the package manager. Making sure that an edit can be applied to the session, is accomplished only with the isInScope call. This call can be conducted at the PackageManager default instance by another class. This class can be activated with ackCreateSession and deactivated with ackLeaveSession.

arminWel replied:
I totally understand your concerns. I'm not satisfied with the architecture aswell. My reasoning was, that the packagemanager does not yet have an association to the client, which would be needed for that. Moreover our architecture tends to give every class an association to every other class, which is kind of ugly. I believe in order to solve this problem we should have a discussion about the complete structure of our architecture (I've to create an class diagramm for our presentation anyway, so maybe that would be a nice starting point for the dicussion, I guess). I'd be glad, if we could do that in the near future(maybe even tomorrow?).

simonstadlinger commented on packages/CollabCodeEdit-Client.package/CCESession.class/instance/broadcastAddedOrModifiedMethod.in..st:
```diff
@@ -0,0 +1,4 @@
+request-sending
+broadcastAddedOrModifiedMethod: aMethodSource in: aClass
```
Strictly speaking this method is not a broadcast. the CCEClient singleton sends a single message to the server which then broadcasts it to all other clients. Because of that, this naming can be confusing when reading all other client side implemented sending methods, where no broadcast is used.

I would therefore suggest to rename the broadcast part of all methods to `sendAddedOrModifiedMethods` etc.

arminWel replied:
Agreed. Still I would like to first have the aforementioned discussion about our architecture before I change that.

simonstadlinger commented on packages/CollabCodeEdit-Core.package/CCEEdit.class/instance/securelyApplyIn..st:
```diff
@@ -0,0 +1,8 @@
+			aCCESession packageManager ignoreEdit: self.
+			self apply 
+			aCCESession packageManager stopIgnoringEdit: self ]
+		ifFalse: [ self error: 'CCEEdit to be applied is not in Scope' ]
```
very nice architecture!

simonstadlinger commented on packages/CollabCodeEdit-Core.package/CCEMessage.class/class/recieveEdit..st:
```diff
@@ -0,0 +1,4 @@
+server-Message creation
+recieveEdit: aCCEEdit
+	^ self new: (CCEMessage serverMsgTypes at: #recieveEdit:from:) 
+			with: aCCEEdit
```
There are no senders of receiveEdit. Therefore i would suggest to delete the method. its functionality is already covered in forewardEditMessage

simonstadlinger replied:
which should be renamed anyway:D

simonstadlinger commented on packages/CollabCodeEdit-Core.package/CCEMessage.class/class/forwardEditMessage..st:
```diff
@@ -1,4 +1,4 @@
 server-Message creation
-forwardEditMessage: aSerializeddCCEEdit
+forwardEditMessage: aCCEEdit
```
```suggestion
forwardEdit: aCCEEdit
```
to stay within the naming convention of CCEMessage. Because every class call lookes like
`CCEMessage ackJoinSession:` etc. it looks a bit odd to append an extra `Message` to every class method.

ADimeo replied:
When this method was named there was an assumption that the Server wouldn't be able to understand the edit, and never actually touch the edit. Within the old assumption the name is valid, since it implies the servers lack of CCEEdit interpretation, just forwarding the message instead.

That assumption has since changed, so a renaming would be in order.

simonstadlinger commented on packages/CollabCodeEdit-Core.package/CCEMessage.class/class/sendEdit..st:
```diff
@@ -1,4 +1,6 @@
-			with: (STON toString: aCCEEdit)
\ No newline at end of file
+	^ self new 
+			type: (CCEMessage clientMsgTypes at: #sendEdit:from:); 
```
is there a reason you separate the new call from the setter methods?

arminWel replied:
That is how cascades should look like.

simonstadlinger commented on packages/CollabCodeEdit-Tests.package/CCEPackageManagerTests.class/instance/setUp.st:
```diff
@@ -1,5 +1,6 @@
+	super setUp.
+	session := CCESessionMock new.
+	session scope: self sessionScope.
+	packageManager listenForChangesFor: session
```
I would suggest to not activate the Change listening until the ackCreateSessionMessage arrives

simonstadlinger commented on packages/CollabCodeEdit-Tests.package/CCEClientMockTest.class/instance/testHasEditOfType.st:
```diff
@@ -0,0 +1,9 @@
+			
+	self assert: (clientMock hasEditOfType: CCEMethodAddedOrModifiedEdit).
+	self deny: (clientMock hasEditOfType: CCEMethodRecategorizedEdit) 
+	
```
Why is there no `self deny: (clientMock hasEditOfType: CCEMethodRemovedEdit)`?

arminWel replied:
In the future there will be even more CCEEdit classes. This serves merely as a sample type, which should not be in the clientMock.

simonstadlinger replied:
Makes totally sens now, leaf it as it is

simonstadlinger commented on packages/CollabCodeEdit-Tests.package/CCEEditTest.class/instance/testRemovedMethodCanBeRemoved.st:
```diff
@@ -0,0 +1,6 @@
+	self compileTestMethod.
+	self assert: (testClass canUnderstand: testMethod).
+	(CCEMethodRemovedEdit newWith: testMethod in: testClass) apply.
+	self deny: (testClass canUnderstand: testMethod)
```
I would suggest to also add a test for a new method. So that all three szenarios are tested

arminWel replied:
Do you search for `CCEEditTest>>testAddedMethodWillBeApplied`?

simonstadlinger replied:
Okay found it. within this method i would add a line before everythink else to deny that the class understands the method.

arminWel replied:
I don't see why this is necessary. If this would be the case it should not alter the behaviour of the test.

simonstadlinger commented on packages/CollabCodeEdit-Tests.package/CCESessionMock.class/instance/broadcastAddedOrModifiedMethod.in..st:
```diff
@@ -0,0 +1,3 @@
+request-sending
+broadcastAddedOrModifiedMethod: aMethodSource in: aClass
+	self messages at: (aClass name, #AddedOrModified) put: aMethodSource
```
Very cool and simple idea to mock the session

simonstadlinger commented on packages/CollabCodeEdit-Tests.package/CCESessionTest.class/instance/testBroadcastAddedOrChangedMethodSendsCCEMessageWithCCEEdit.st:
```diff
@@ -0,0 +1,4 @@
+tests
+testBroadcastAddedOrChangedMethodSendsCCEMessageWithCCEEdit
+	session broadcastAddedOrModifiedMethod: #test in: CCETestDummy.
+	self assert: (clientMock hasEditOfType: CCEMethodAddedOrModifiedEdit)
```
i would suggest to clear the mock messages after every test, because then it can be ensured that every test finds a empty interface which makes the tests more reliable

simonstadlinger reviewed (dismissed):
Very carefully implemented feature, thank you for that. 
But there is a architectural issue i would like to discuss on. You can find my concerns in the comments.

arminWel committed:
Give the packageManager complete responsibility over sending the edits

Before this commit, the session merely forwarded the CCEEdit to the client. This type of architecture was criticised in the review.

arminWel committed:
Clean up after architectural change to fix broken tests

PaulBrachmann commented on packages/CollabCodeEdit-Client.package/CCEClient.class/instance/sendMessage..st:
```diff
@@ -0,0 +1,4 @@
+sending
+sendMessage: aCCEMessage
+	Transcript show: aCCEMessage serialize.
```
Debug code should not be pulled onto develop

arminWel replied:
If i remember correctly, we already got `Transcript show:` in the develop, but your absolutely right.

PaulBrachmann commented on packages/CollabCodeEdit-Client.package/CCEModel.class/instance/initialize.st:
```diff
@@ -1,11 +1,12 @@
+		client: (CCEClient default owner: self);
+		session: (CCESession default client: self client).
+	self session 
+				addStatusSubscriber: self;
```
This shows as some really weird indentation. It might be just GitHub, but please check for consistency.

PaulBrachmann commented on packages/CollabCodeEdit-Client.package/CCEModel.class/instance/noSessionIsLoaded.st:
```diff
@@ -1,3 +1,3 @@
-as yet unclassified
+testing
 noSessionIsLoaded
 	^ self session getStatus = CCESessionStatus noRemoteSession.
```
While you're at it, please remove the trailing point.

PaulBrachmann commented on packages/CollabCodeEdit-Server.package/CCERemoteClient.class/instance/sendMessageToSession..st:
```diff
@@ -0,0 +1,4 @@
+sending
+sendMessageToSession: aCCEMessage
+	self session broadcastMessage: aCCEMessage without: self
+	
```
Remove the trailing empty line

PaulBrachmann commented on packages/CollabCodeEdit-Server.package/CCEServer.class/instance/CCEcreateSessionWith.from..st:
```diff
@@ -7,5 +7,5 @@ CCEcreateSessionWith: anOrderedDictionary from: aClient
 
-	aClient sendMsg: (CCEMessage ackCreateSession: sessionToken)
+	aClient sendMessage: (CCEMessage ackCreateSession: sessionToken)
 
```
Remove the trailing empty line

PaulBrachmann commented on packages/CollabCodeEdit-Server.package/CCEServer.class/instance/CCEleaveSessionWith.from..st:
```diff
@@ -4,7 +4,7 @@ CCEleaveSessionWith: anOrderedDictionary from: aClient
 		ifFalse: [
-			aClient sendMsg: (CCEMessage ackLeaveSessionWithError: {#error -> 'Invalid session token'}) ]
+			aClient sendMessage: (CCEMessage ackLeaveSessionWithError: {#error -> 'Invalid session token'}) ]
 	
```
Remove the trailing empty line

PaulBrachmann commented on packages/CollabCodeEdit-Tests.package/CCEClientManagerTest.class/instance/setUp.st:
```diff
@@ -0,0 +1,6 @@
+	clients := (OrderedCollection newFrom: { CCEClientMock new . CCEClientMock new. CCEClientMock new. }).
+	clientManager :=  CCEClientManager new.
+	clients do: [ :client | clientManager addClient: client ]
+	
```
Remove the trailing empty line

PaulBrachmann commented on packages/CollabCodeEdit-Tests.package/CCEClientManagerTest.class/instance/testBroadcastMessageSendsToAll.st:
```diff
@@ -0,0 +1,10 @@
+					
+	clientManager broadcastMessage: aCCEMessage.
+	
+	clients do: [ :client | self assert: (client hasMessage: aCCEMessage)]
```
```suggestion
	clients do: [ :client | self assert: (client hasMessage: aCCEMessage) ]
```

PaulBrachmann commented on packages/CollabCodeEdit-Tests.package/CCEClientManagerTest.class/instance/setUp.st:
```diff
@@ -0,0 +1,6 @@
+running
+setUp
+	clients := (OrderedCollection newFrom: { CCEClientMock new . CCEClientMock new. CCEClientMock new. }).
```
```suggestion
	clients := (OrderedCollection newFrom: { CCEClientMock new . CCEClientMock new . CCEClientMock new }).
```

PaulBrachmann commented on packages/CollabCodeEdit-Tests.package/CCEClientManagerTest.class/instance/testBroadcastMessageWithoutSendsToAllExceptOne.st:
```diff
@@ -0,0 +1,11 @@
+	clientManager broadcastMessage: aCCEMessage without: clients first.
+	
+	clients do: [ :client | self assert: (client hasMessage: aCCEMessage)] without: clients first.
+	self assert: clients first hasNoMessage.
```
Remove the trailing point

PaulBrachmann commented on packages/CollabCodeEdit-Tests.package/CCESocketMock.class/instance/hasMessage..st:
```diff
@@ -1,4 +1,4 @@
-	self messages keysDo: [:key | key = aString ifTrue: [^ true]].
+testing
+hasMessage: aMessage
+	self messages keysDo: [:key | key = aMessage ifTrue: [^ true]].
```
```suggestion
	self messages keysDo: [ :key | key = aMessage ifTrue: [ ^ true ] ].
```

PaulBrachmann reviewed (changes requested):
This is some beautiful code with minor styling inconsistencies that should be fixable in no time

arminWel committed:
Fix style mistakes found in @PaulBrachmann 's review

PaulBrachmann commented on packages/CollabCodeEdit-Tests.package/CCESocketMock.class/instance/hasMessage..st:
```diff
@@ -1,4 +1,4 @@
-	self messages keysDo: [:key | key = aString ifTrue: [^ true]].
+testing
+hasMessage: aMessage
+	self messages keysDo: [ :key | key = aMessage ifTrue: [^ true]].
```
```suggestion
	self messages keysDo: [ :key | key = aMessage ifTrue: [ ^ true ] ].
```

PaulBrachmann reviewed (changes requested):
There still are plenty of comments from the last review that are not yet addressed. Please consider them before I approve the PR.

arminWel committed:
Fix even more style mistakes
